### PR TITLE
graphql: cache GraphQL validation

### PR DESCRIPTION
**Pull request by kamilkisiela**
I ran the same query ~55 times and got these results:

```
No caching (58 samples)

p95:      47.75µs
p99:      73.83µs
average:  24.85µs
median:   20.13µs

Caching (56 samples)
    
p95:       4.63µs
p99:       4.92µs
average:   2.08µs
median:    1.67µs
```

Part of https://github.com/graphprotocol/graph-node/issues/3077

### CODE REVIEW

File: graphql/src/execution/query.rs [hunk 1]
```diff
@@ -150,8 +156,16 @@ impl Query {
...
+        let mut validation_cache = GRAPHQL_VALIDATION_CACHE
+            .lock()
+            .unwrap_or_else(PoisonError::into_inner);
+        let validation_errors = validation_cache.entry(query.shape_hash).or_insert_with(|| {
+            validate(
+                &schema.document(),
+                &query.document,
+                &GRAPHQL_VALIDATION_PLAN,
+            )
+        });
```
Comment (leoyvens) [inline graphql/src/execution/query.rs:168]: It seems the lock is held for a larger scope than necessary, maybe you can do it all in one line:
```suggestion
        GRAPHQL_VALIDATION_CACHE
            .lock()
            .unwrap_or_else(PoisonError::into_inner)
            .entry(query.shape_hash).or_insert_with(|| {
            validate(
                &schema.document(),
                &query.document,
                &GRAPHQL_VALIDATION_PLAN,
            )
        });
```

**Review by leoyvens — CHANGES_REQUESTED** [review]

**Comment by kamilkisiela** [comment]
Applied suggested changes, ran `cargo fmt` and squashed commits

**Comment by leoyvens** [comment]
Still doesn't build.

File: graphql/src/execution/query.rs (hunk 1 above)
Comment (kamilkisiela) [reply graphql/src/execution/query.rs]: @leoyvens not sure how to do it in one line and assign the result to `let validation_errors`

**Comment by leoyvens** [comment]
Lock lifetimes are tricky. I've built on your changes so that the lock is only held when reading or inserting to the cache, but importantly not held when actually validating the query. As you had noticed this requires a bit of cloning.

**Comment by leoyvens** [comment]
@kamilkisiela could you review and re-run the benchmarks?

**Comment by kamilkisiela** [comment]
@leoyvens thank you. I re-run the benchmarks and it looks faster now.

**Review by leoyvens — APPROVED** [review]

**Pull request merged**